test/Gen.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE DataKinds, DeriveFunctor, DeriveGeneric, FlexibleInstances, FunctionalDependencies, GADTs, GeneralizedNewtypeDeriving, LambdaCase, PatternSynonyms, PolyKinds, RankNTypes, ScopedTypeVariables, StandaloneDeriving, TypeApplications, TypeOperators, UndecidableInstances, ViewPatterns #-}
+module Gen
+( module Control.Carrier.Pure
+  -- * Polymorphic generation & instantiation
+, m
+, genT
+, a
+, b
+, e
+, r
+, s
+, w
+, T(..)
+, A
+, B
+, E
+, R
+, S
+, W
+  -- * Generation
+, Rec(..)
+, forall
+  -- * Showing generated values
+, With(getWith)
+, showing
+, showingFn
+, atom
+, liftWith
+, liftWith2
+  -- * Pattern synonyms
+, pattern With
+, pattern Fn
+, pattern FnWith
+  -- * Re-exports
+, Gen
+, (===)
+, (/==)
+, choice
+, subterm
+, subtermM
+, subterm2
+, Arg
+, Vary
+, fn
+, apply
+) where
+
+import Control.Carrier.Pure
+import Data.Function (on)
+import Data.Functor.Classes (showsUnaryWith)
+import Data.Proxy
+import GHC.Stack
+import GHC.TypeLits
+import Hedgehog
+import Hedgehog.Function hiding (R, S)
+import Hedgehog.Gen
+import Hedgehog.Range
+
+-- | A generator for computations, given a higher-order generator for effectful operations, & a generator for results.
+m
+  :: forall m a
+  .  (Applicative m, Show a)
+  => (forall a . Show a => (forall a . Show a => Gen a -> Gen (With (m a))) -> Gen a -> Gen (With (m a))) -- ^ A higher-order generator producing operations using any effects in @m@.
+  -> Gen a                                                                                      -- ^ A generator for results.
+  -> Gen (With (m a))                                                                           -- ^ A generator producing computations, wrapped in 'With' for convenience.
+m with = go where
+  go :: forall a . Show a => Gen a -> Gen (With (m a))
+  go a = recursive choice [(liftWith "pure" pure . showing) <$> a] [with go a]
+
+
+genT :: Gen (T a)
+genT = T <$> integral (linear 0 100)
+
+newtype T a = T { unT :: Integer }
+  deriving (Enum, Eq, Generic, Num, Ord, Real, Vary)
+
+instance Arg (T a)
+
+instance KnownSymbol s => Show (T s) where
+  showsPrec d = showsUnaryWith showsPrec (symbolVal (Proxy @s)) d . unT
```

Comment:
`Test.QuickCheck.Poly` offered `A`, `B`, and `C` newtypes for convenient testing of polymorphic properties. I figured I could improve on them a bit.

1. `Arg` and `Vary` are used by `hedgehog-fn` to generate functions.

2. `T` has a phantom, poly-kinded type parameter enabling us to reuse a single type for multiple different “roles”—e.g. if a property requires handlers, continuations, variables, and actions, the generators for all of these can involve the single polymorphic generator for `T` at some level.

3. The `Show` instance for `T` uses the type parameter for the constructor name, enabling us to show a value of `T { unT = 0 } :: T "A"` much more succinctly _and_ readably as `A 0`. This makes it obvious at a glance which role a value in a test failure was filling, without having to sort out which `forAll` in the failure message corresponds to which variable in the property.